### PR TITLE
DAOS-11180 dtx: trigger DTX reindex before DTX resync

### PR DESCRIPTION
**Pull request by Nasf-Fan**
master-commit: 8daa893750b6ca5d4a34c8ff080471a565f0d5f6

During DTX resync, the new DTX leader will send DTX_CHECK RPC
to other alive DTX participant(s) to query related DTX status.
If related DTX entry does not exist on the non-leader and DTX
reindex on such non-leader is not completed yet, then the non
leader will reply to the new leader with -DER_INPROGRESS that
will cause the new leader to retry the DTX_CHECK RPC sometime
later. But if DTX reindex is not triggerred on the non-leader,
then the new leader will retry for ever. So the patch triggers
DTX reindex before DTX resync scanning.

The patch also adjusts the logic of filter DTX entries when
DTX resync to avoid skipping some DTX entries.

It also contains the fixes from PR#9938:

If the DTX_CHECK RPC handler hits -DER_INPROGRESS failure on some
target, then it means that the committed DTX entries on such target
are not re-indexed. Under such case, we need to trigger DTX re-index
on the target for subsequent (retry) DTX_CHECK RPC. Otherwise, it may
cause more DTX_CHECK RPCs to be retried (for ever), as to related DTX
resync cannot complete on some DTX leaders.

Signed-off-by: Fan Yong <fan.yong@intel.com>

### CODE REVIEW

**Review by jolivier23 — DISMISSED** [review]

**Comment by jolivier23** [comment]
waiting for merge approval still

**Comment by jolivier23** [comment]
I landed another patch this morning for this ticket.  Is the plan to combine into one backport?

**Comment by Nasf-Fan** [comment]
> I landed another patch this morning for this ticket. Is the plan to combine into one backport?

Yes, it contains both.

**Review by jolivier23 — APPROVED** [review]

**Comment by jolivier23** [comment]
@wangdi1 can you please review since it's unclean-cherry-pick ?

File: src/dtx/dtx_common.c
Comment (wangdi1) [inline src/dtx/dtx_common.c:1515]: Just curious, does it need wait until dtx reindex ULT finish?

**Review by wangdi1 — APPROVED** [review]

File: src/dtx/dtx_common.c
Comment (Nasf-Fan) [reply src/dtx/dtx_common.c:1515]: I do not think so. The DTX reindex ULT holds reference on the container, but does not hold open count. stop_dtx_reindex_ult() is usually called when close the container. It is normal the we still have some references on the container even if the container is closed.

**Pull request merged**
